Maven plugin always downloads .avsc file extension even registred non Avro schema

I registered a protocol buffer schema via maven plugin and all registered fine.
When I tried to download it in the maven plugin in other project, the file created with the .avsc extension

@alesj , @EricWittmann it's very strange, but adding new maven-plugin capabilities and calling simple 
`ArtifactMetaData artifactMetaData = getClient().getArtifactMetaData(id);` got following error
```
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: org.glassfish.jersey.internal.RuntimeDelegateImpl
	at javax.ws.rs.ext.RuntimeDelegate.findDelegate(RuntimeDelegate.java:129)
	at javax.ws.rs.ext.RuntimeDelegate.getInstance(RuntimeDelegate.java:96)
	at javax.ws.rs.core.MediaType.valueOf(MediaType.java:172)
	at io.apicurio.registry.client.RegistryRestClientImpl.parseResponseBody(RegistryRestClientImpl.java:325)
	at io.apicurio.registry.client.RegistryRestClientImpl.getLatestArtifact(RegistryRestClientImpl.java:133)
	at io.apicurio.registry.maven.DownloadRegistryMojo.executeInternal(DownloadRegistryMojo.java:122)
```

--- a/utils/maven-plugin/pom.xml
+++ b/utils/maven-plugin/pom.xml
@@ -15,10 +15,20 @@
     <packaging>maven-plugin</packaging>
     <name>apicurio-registry-maven-plugin</name>
 
+    <properties>
+        <resteasy-client.version>4.5.6.Final</resteasy-client.version>
+    </properties>
+
     <dependencies>
         <dependency>
             <groupId>io.apicurio</groupId>
             <artifactId>apicurio-registry-rest-client</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-client</artifactId>
+            <version>${resteasy-client.version}</version>
         </dependency>
 
         <dependency>

--- a/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/DownloadRegistryMojo.java
+++ b/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/DownloadRegistryMojo.java
@@ -17,6 +17,8 @@
 
 package io.apicurio.registry.maven;
 
+import io.apicurio.registry.rest.beans.ArtifactMetaData;
+import io.apicurio.registry.types.ArtifactExtensionType;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
@@ -54,9 +56,9 @@ public class DownloadRegistryMojo extends AbstractRegistryMojo {
     Map<String, Integer> versions = new LinkedHashMap<>();
 
     /**
-     * Default artifact file extension to use when writing the artifact to file-system.
+     * Overwrite artifact file extension to use when writing the artifact to the filesystem.
      */
-    @Parameter(defaultValue = ".avsc")
+    @Parameter
     String artifactExtension;
 
     /**
@@ -96,6 +98,18 @@ public class DownloadRegistryMojo extends AbstractRegistryMojo {
 
         for (String id : ids) {
             String ext = artifactExtensions.getOrDefault(id, artifactExtension);
+            // Explicit file extension is not defined, getting it from the metadata
+            if (ext == null || ext.equals("")) {
+                try {
+                    ArtifactMetaData artifactMetaData = getClient().getArtifactMetaData(id);
+                    ext = ".".concat(ArtifactExtensionType.fromArtifactType(artifactMetaData.getType()).toString());
+                } catch (Exception ex) {
+                    throw new MojoExecutionException(
+                            String.format("Exception thrown while getting artifact [%s] metadata", id),
+                            ex
+                    );
+                }
+            }
             String fileName = String.format("%s%s", id, ext);
             File outputFile = new File(outputDirectory, fileName);
 

--- a/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/RegisterRegistryMojo.java
+++ b/utils/maven-plugin/src/main/java/io/apicurio/registry/maven/RegisterRegistryMojo.java
@@ -17,7 +17,6 @@
 
 package io.apicurio.registry.maven;
 
-import java.io.IOException;
 import java.io.InputStream;
 import java.util.LinkedHashMap;
 import java.util.Map;
@@ -42,7 +41,7 @@ public class RegisterRegistryMojo extends ContentRegistryMojo {
      */
     Map<String, Integer> artifactVersions;
 
-    public ArtifactMetaData register(String artifactId, ArtifactType artifactType, StreamHandle handle) throws IOException {
+    public ArtifactMetaData register(String artifactId, ArtifactType artifactType, StreamHandle handle) {
         try {
             try (InputStream stream = handle.stream()) {
                 return getClient().createArtifact(artifactType, artifactId, IfExistsType.RETURN_OR_UPDATE, stream);

--- a/utils/maven-plugin/src/main/java/io/apicurio/registry/types/ArtifactExtensionType.java
+++ b/utils/maven-plugin/src/main/java/io/apicurio/registry/types/ArtifactExtensionType.java
@@ -1,0 +1,47 @@
+package io.apicurio.registry.types;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public enum ArtifactExtensionType {
+    AVRO(ArtifactType.AVRO, "avsc"),
+    PROTOBUF(ArtifactType.PROTOBUF, "proto"),
+    PROTOBUF_FD(ArtifactType.PROTOBUF_FD, "json"),
+    JSON(ArtifactType.JSON, "json"),
+    OPENAPI(ArtifactType.OPENAPI, "json"),
+    ASYNCAPI(ArtifactType.ASYNCAPI, "json"),
+    GRAPHQL(ArtifactType.GRAPHQL, "graphql"),
+    KCONNECT(ArtifactType.KCONNECT, "json"),
+    WSDL(ArtifactType.WSDL, "wsdl"),
+    XSD(ArtifactType.XSD, "xsd"),
+    XML(ArtifactType.XML, "xml");
+
+    private final ArtifactType artifactType;
+    private final String artifactExtension;
+    private final static Map<ArtifactType, ArtifactExtensionType> CONSTANTS = new HashMap<>();
+
+    static {
+        for (ArtifactExtensionType c: values()) {
+            CONSTANTS.put(c.artifactType, c);
+        }
+    }
+
+    ArtifactExtensionType(ArtifactType artifactType, String artifactExtension) {
+        this.artifactType = artifactType;
+        this.artifactExtension = artifactExtension;
+    }
+
+    @Override
+    public String toString() {
+        return this.artifactExtension;
+    }
+
+    public static ArtifactExtensionType fromArtifactType(ArtifactType artifactType) {
+        ArtifactExtensionType constant = CONSTANTS.get(artifactType);
+        if (constant == null) {
+            throw new IllegalArgumentException(artifactType.toString());
+        } else {
+            return constant;
+        }
+    }
+}

--- a/utils/maven-plugin/src/test/resources/test-build/pom.xml
+++ b/utils/maven-plugin/src/test/resources/test-build/pom.xml
@@ -16,7 +16,7 @@
 		<maven.compiler.source>1.8</maven.compiler.source>
 		<maven.compiler.target>1.8</maven.compiler.target>
 		
-		<apicurio.version>1.2.2-SNAPSHOT</apicurio.version>
+		<apicurio.version>1.3.1-SNAPSHOT</apicurio.version>
 	</properties>
 
 	<build>

